include/ensmallen_bits/nelder_mead/nelder_mead_impl.hpp

```diff
@@ -0,0 +1,241 @@
+  {
+    const MatType M = arma::mean(simplex.cols(order.subvec(0, dim - 1)), 1);
+    const MatType xref = (1.0 + alpha) * M - alpha * simplex.col(order(dim));
+    const ElemType fref = function.Evaluate(xref);
+    Callback::Evaluate(*this, function, xref, fref, callbacks...);
+
+    if (fref < fx(order(0)))
+    {
+      // Compute expansion.
+      const MatType xexp = (1.0 + alpha * beta) * M - alpha * beta *
+          simplex.col(order(dim));
+      const ElemType fexp = function.Evaluate(xexp);
+      Callback::Evaluate(*this, function, xexp, fexp, callbacks...);
+
+      // Shift values by one to the right and swap the last with the first.
+      const size_t fLowest = order(dim);
+      for (size_t d = dim; d > 0; --d)
+        order(d) = order(d - 1);
+      order(0) = fLowest;
+
+      if (fexp < fref)
+      {
+        simplex.col(order(0)) = xexp;
+        fx(order(0)) = fexp;
+      }
+      else
+      {
+        simplex.col(order(0)) = xref;
+        fx(order(0)) = fref;
+      }
+    }
+    else if (fref < fx(order(dim - 1)))
+    {
+      // Accept reflection point.
+      const size_t s = ShiftOrder(order, fx, fref);
+      simplex.col(order(s)) = xref;
+      fx(order(s)) = fref;
+    }
+    else
+    {
+      if (fref < fx(order(dim)))
+      {
+        // Outside contraction.
+        const MatType xoc = (1.0 + alpha * gamma) * M - alpha * gamma *
+            simplex.col(order(dim));
+        const ElemType foc = function.Evaluate(xoc);
+        Callback::Evaluate(*this, function, xoc, foc, callbacks...);
+
+        if (foc <= fref)
+        {
+          const size_t s = ShiftOrder(order, fx, fref);
+          simplex.col(order(s)) = xoc;
+          fx(order(s)) = foc;
+        }
+        else
+        {
+          shrink = true;
+        }
+      }
+      else
+      {
+        // Inside contraction.
+        const MatType xic = (1.0 - gamma) * M + gamma *
+            simplex.col(order(dim));
+        const ElemType fic = function.Evaluate(xic);
+        Callback::Evaluate(*this, function, xic, fic, callbacks...);
+
+        if (fic < fx(order(dim)))
+        {
+          const size_t s = ShiftOrder(order, fx, fic);
+          simplex.col(order(s)) = xic;
+          fx(order(s)) = fic;
+        }
+        else
+        {
+          shrink = true;
+        }
+      }
+    }
+
+    if (shrink)
+    {
+      for (size_t d = 1; d < dim; ++d)
+      {
+        simplex.col(order(d)) = simplex.col(order(0)) + delta *
+            (simplex.col(order(d)) - simplex.col(order(0)));
+        fx(order(d)) = function.Evaluate(simplex.col(order(d)));
+
+        Callback::Evaluate(*this, function, simplex.col(order(d)),
+            fx(order(d)), callbacks...);
+      }
+
+      // Sorting can be slow, however there is overwhelming evidence that shrink
+      // transformations almost never happen in practice. See "Efficient
+      // Implementation of the Nelder–Mead Search Algorithm" by S. Singer and S.
+      // Singer.
```

Comment:
I thought for sure this had to be a typo, but no, there are actually *two* S. Singers, and they collaborated on this. :smile: